c/tskit/trees.c

```diff
@@ -7814,75 +7835,455 @@ remove_unextended(edge_list_t **head, edge_list_t **tail)
+    bool *not_sample;
+    double *near_side, *far_side;
+    edge_list_t *edges_out_head, *edges_out_tail;
+    edge_list_t *edges_in_head, *edges_in_tail;
+    tsk_blkalloc_t edge_list_heap;
+} haplotype_extender_t;
+
+static int
+haplotype_extender_init(haplotype_extender_t *self, const tsk_treeseq_t *ts,
+    int direction, tsk_edge_table_t *edges)
+{
+    int ret = 0;
+    tsk_id_t tj;
+    tsk_size_t num_nodes = tsk_treeseq_get_num_nodes(ts);
+
+    tsk_memset(self, 0, sizeof(haplotype_extender_t));
+
+    self->ts = ts;
+    self->edges = edges;
+    ret = tsk_edge_table_copy(&ts->tables->edges, self->edges, TSK_NO_INIT);
     if (ret != 0) {
         goto out;
     }
-    ret = tsk_edge_table_copy(&tables->edges, edges, TSK_NO_INIT);
+
+    self->direction = direction;
+    if (direction == TSK_DIR_FORWARD) {
+        self->near_side = self->edges->left;
+        self->far_side = self->edges->right;
+    } else {
+        self->near_side = self->edges->right;
+        self->far_side = self->edges->left;
+    }
+
+    self->edges_in_head = NULL;
+    self->edges_in_tail = NULL;
+    self->edges_out_head = NULL;
+    self->edges_out_tail = NULL;
+
+    tsk_memset(&self->edge_list_heap, 0, sizeof(self->edge_list_heap));
+    ret = tsk_blkalloc_init(&self->edge_list_heap, 8192);
     if (ret != 0) {
         goto out;
     }
 
-    for (tj = 0; tj < (tsk_id_t) tables->nodes.num_rows; tj++) {
-        not_sample[tj] = ((tables->nodes.flags[tj] & TSK_NODE_IS_SAMPLE) == 0);
+    self->last_degree = tsk_calloc(num_nodes, sizeof(*self->last_degree));
+    self->next_degree = tsk_calloc(num_nodes, sizeof(*self->next_degree));
+    self->last_nodes_edge = tsk_malloc(num_nodes * sizeof(*self->last_nodes_edge));
+    self->next_nodes_edge = tsk_malloc(num_nodes * sizeof(*self->next_nodes_edge));
+    self->parent_out = tsk_malloc(num_nodes * sizeof(*self->parent_out));
+    self->parent_in = tsk_malloc(num_nodes * sizeof(*self->parent_in));
+    self->not_sample = tsk_malloc(num_nodes * sizeof(*self->not_sample));
+
+    if (self->last_degree == NULL || self->next_degree == NULL
+        || self->last_nodes_edge == NULL || self->next_nodes_edge == NULL
+        || self->parent_out == NULL || self->parent_in == NULL
+        || self->not_sample == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    tsk_memset(self->last_nodes_edge, 0xff, num_nodes * sizeof(*self->last_nodes_edge));
+    tsk_memset(self->next_nodes_edge, 0xff, num_nodes * sizeof(*self->next_nodes_edge));
+    tsk_memset(self->parent_out, 0xff, num_nodes * sizeof(*self->parent_out));
+    tsk_memset(self->parent_in, 0xff, num_nodes * sizeof(*self->parent_in));
+
+    for (tj = 0; tj < (tsk_id_t) num_nodes; tj++) {
+        self->not_sample[tj] = ((ts->tables->nodes.flags[tj] & TSK_NODE_IS_SAMPLE) == 0);
     }
 
-    if (forwards) {
-        near_side = edges->left;
-        far_side = edges->right;
+out:
+    return ret;
+}
+
+// static void
```

Comment:
I tend to make the ``print_state`` a part of the public interface so that it can be tested and it stays up to date. I guess the struct here is not exposed in the header so there's no way to access it. We worked around this for simplify by adding a ``TSK_DEBUG`` option [here](https://github.com/tskit-dev/tskit/blob/024e0425139efcf8b6c30229699910a5757a71fa/c/tskit/tables.c#L12106)